File: programacion3/programacion3-2/Backtracking/N-Reinas/N-Reinas.exs
defmodule NReinas do
  def main do
    n =
      "Ingrese el número de reinas: "
      |> Util.ingresar(:entero)

    IO.puts("\n=== PROBLEMA DE LAS #{n} REINAS ===")
    resolver(n)
  end

  def resolver(n) do
    colocar_reinas(n, 0, [])
  end

  defp colocar_reinas(n, fila, tablero) when fila == n do
    IO.puts("\nSolución encontrada:")
    mostrar_tablero(Enum.reverse(tablero), n)
  end

  defp colocar_reinas(n, fila, tablero) do
    for col <- 0..(n - 1),
        es_valido?(fila, col, tablero),
        do: colocar_reinas(n, fila + 1, [{fila, col} | tablero])
  end

  defp es_valido?(fila, col, tablero) do
    Enum.all?(tablero, fn {f, c} ->
      c != col and abs(c - col) != abs(f - fila)
    end)
  end

  defp mostrar_tablero(tablero, n) do
    for fila <- 0..(n - 1) do
      for col <- 0..(n - 1) do
        if Enum.member?(tablero, {fila, col}), do: IO.write(" Q "), else: IO.write(" x ")
      end
      IO.puts("")
    end
  end
end

NReinas.main()

File: programacion3/programacion3-2/Backtracking/N-Reinas/NReinas.java
import java.util.Scanner;

public class NReinas {
    static int N;

    public static void main(String[] args) {
        try (Scanner sc = new Scanner(System.in)) {
            System.out.print("Ingrese el número de reinas: ");
            N = sc.nextInt();
            int[] tablero = new int[N];
            resolver(tablero, 0);
        }
    }

    static void resolver(int[] tablero, int fila) {
        if (fila == N) {
            System.out.println("\nSolución encontrada:");
            mostrar(tablero);
            return;
        }
        for (int col = 0; col < N; col++) {
            if (esValido(tablero, fila, col)) {
                tablero[fila] = col;
                resolver(tablero, fila + 1);
            }
        }
    }

    static boolean esValido(int[] tablero, int fila, int col) {
        for (int i = 0; i < fila; i++) {
            if (tablero[i] == col || Math.abs(tablero[i] - col) == Math.abs(i - fila))
                return false;
        }
        return true;
    }

    static void mostrar(int[] tablero) {
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                System.out.print(tablero[i] == j ? " Q " : " x ");
            }
            System.out.println();
        }
    }
}
